### ISSUE TITLE
Panic!

### ISSUE THREAD

**Issue by nuxeh** [nuxeh/url-bot-rs#79]
```
RESOLVE https://github.com/pazz/alot/compare/418fa67867af^...48626503caa3
thread 'reqwest-internal-sync-runtime' panicked at 'a parsed Url should always be a valid Uri: InvalidUri(InvalidUriChar)', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread 'main' panicked at 'event loop thread panicked', /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/client.rs:482:13
```

**Comment by polyzen** [comment]
![](https://eured.files.wordpress.com/2010/10/oh-noes-everybody-panic.gif)

**Comment by nuxeh** [comment]
Full backtrace:
```
Using in-memory database
RESOLVE https://github.com/pazz/alot/compare/418fa67867af^...48626503caa3
thread 'reqwest-internal-sync-runtime' panicked at 'a parsed Url should always be a valid Uri: InvalidUri(InvalidUriChar)', libcore/result.rs:1009:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:391
   6: rust_begin_unwind
             at libstd/panicking.rs:326
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:77
   8: core::result::unwrap_failed
             at libcore/macros.rs:26
   9: <core::result::Result<T, E>>::expect
             at libcore/result.rs:835
  10: reqwest::into_url::to_uri
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/into_url.rs:44
  11: reqwest::async_impl::client::Client::execute_request
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/async_impl/client.rs:336
  12: reqwest::async_impl::client::Client::execute
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/async_impl/client.rs:313
  13: reqwest::client::ClientHandle::new::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/client.rs:433
  14: <futures::stream::for_each::ForEach<S, F, U> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/stream/for_each.rs:46
  15: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  16: <futures::task_impl::Spawn<T>>::enter::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  17: futures::task_impl::std::set
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/std/mod.rs:78
  18: <futures::task_impl::Spawn<T>>::enter
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  19: <futures::task_impl::Spawn<T>>::poll_fn_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:288
  20: <futures::task_impl::Spawn<T>>::poll_future_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  21: <tokio_current_thread::Entered<'a, P>>::block_on::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:479
  22: <tokio_current_thread::Borrow<'a, U>>::enter::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:779
  23: tokio_current_thread::CurrentRunner::set_spawn
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:816
  24: <tokio_current_thread::Borrow<'a, U>>::enter::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:778
  25: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  26: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  27: <tokio_current_thread::Borrow<'a, U>>::enter
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:776
  28: <tokio_current_thread::Entered<'a, P>>::block_on
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:478
  29: tokio::runtime::current_thread::runtime::Runtime::block_on::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/current_thread/runtime.rs:187
  30: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/current_thread/runtime.rs:228
  31: tokio_executor::global::with_default::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.5/src/global.rs:192
  32: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  33: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  34: tokio_executor::global::with_default
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.5/src/global.rs:162
  35: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/current_thread/runtime.rs:226
  36: tokio_timer::timer::handle::with_default::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.7/src/timer/handle.rs:94
  37: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  38: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  39: tokio_timer::timer::handle::with_default
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.7/src/timer/handle.rs:81
  40: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/current_thread/runtime.rs:219
  41: tokio_timer::clock::clock::with_default::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.7/src/clock/clock.rs:136
  42: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  43: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  44: tokio_timer::clock::clock::with_default
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.7/src/clock/clock.rs:119
  45: tokio::runtime::current_thread::runtime::Runtime::enter::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/current_thread/runtime.rs:218
  46: tokio_reactor::with_default::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.6/src/lib.rs:229
  47: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  48: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  49: tokio_reactor::with_default
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.6/src/lib.rs:212
  50: tokio::runtime::current_thread::runtime::Runtime::enter
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/current_thread/runtime.rs:217
  51: tokio::runtime::current_thread::runtime::Runtime::block_on
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/current_thread/runtime.rs:185
  52: reqwest::client::ClientHandle::new::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/client.rs:441
thread 'main' panicked at 'event loop thread panicked', /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/client.rs:482:13
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:477
   5: std::panicking::begin_panic
             at libstd/panicking.rs:411
   6: reqwest::client::ClientHandle::execute_request::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/client.rs:482
   7: <core::result::Result<T, E>>::map_err
             at libcore/result.rs:530
   8: <futures::future::map_err::MapErr<A, F> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/map_err.rs:34
   9: <futures::future::join::MaybeDone<A>>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/join.rs:153
  10: <futures::future::join::Join<A, B> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/join.rs:75
  11: <futures::future::map::Map<A, F> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/map.rs:30
  12: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  13: <futures::task_impl::Spawn<T>>::enter::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  14: futures::task_impl::std::set
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/std/mod.rs:78
  15: <futures::task_impl::Spawn<T>>::enter
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  16: <futures::task_impl::Spawn<T>>::poll_fn_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:288
  17: <futures::task_impl::Spawn<T>>::poll_future_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  18: reqwest::wait::timeout
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/wait.rs:26
  19: reqwest::client::ClientHandle::execute_request
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/client.rs:487
  20: reqwest::client::Client::execute
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/client.rs:362
  21: reqwest::request::RequestBuilder::send
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.9.2/src/request.rs:484
  22: url_bot::http::resolve_url
             at src/http.rs:23
  23: url_bot::message::handle_message
             at src/message.rs:44
  24: url_bot::main::{{closure}}
             at src/main.rs:103
  25: irc::client::reactor::IrcReactor::register_client_with_handler::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/irc-0.13.6/src/client/reactor.rs:145
  26: <futures::stream::for_each::ForEach<S, F, U> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/stream/for_each.rs:46
  27: <alloc::boxed::Box<F> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/mod.rs:113
  28: <futures::future::join_all::JoinAll<I> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/join_all.rs:100
  29: <futures::future::map::Map<A, F> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/map.rs:30
  30: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  31: <futures::task_impl::Spawn<T>>::enter::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  32: futures::task_impl::std::set
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/std/mod.rs:78
  33: <futures::task_impl::Spawn<T>>::enter
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  34: <futures::task_impl::Spawn<T>>::poll_fn_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:288
  35: <futures::task_impl::Spawn<T>>::poll_future_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  36: tokio_core::reactor::Core::run::{{closure}}::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.17/src/reactor/mod.rs:236
  37: <futures::future::lazy::Lazy<F, R>>::get
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/lazy.rs:64
  38: <futures::future::lazy::Lazy<F, R> as futures::future::Future>::poll
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/future/lazy.rs:82
  39: <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  40: <futures::task_impl::Spawn<T>>::enter::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  41: futures::task_impl::std::set
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/std/mod.rs:78
  42: <futures::task_impl::Spawn<T>>::enter
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:396
  43: <futures::task_impl::Spawn<T>>::poll_fn_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:288
  44: <futures::task_impl::Spawn<T>>::poll_future_notify
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.25/src/task_impl/mod.rs:326
  45: <tokio_current_thread::Entered<'a, P>>::block_on::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:479
  46: <tokio_current_thread::Borrow<'a, U>>::enter::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:779
  47: tokio_current_thread::CurrentRunner::set_spawn
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:816
  48: <tokio_current_thread::Borrow<'a, U>>::enter::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:778
  49: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  50: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  51: <tokio_current_thread::Borrow<'a, U>>::enter
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:776
  52: <tokio_current_thread::Entered<'a, P>>::block_on
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-current-thread-0.1.3/src/lib.rs:478
  53: tokio_core::reactor::Core::run::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.17/src/reactor/mod.rs:234
  54: tokio_timer::timer::handle::with_default::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.7/src/timer/handle.rs:94
  55: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  56: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  57: tokio_timer::timer::handle::with_default
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.7/src/timer/handle.rs:81
  58: tokio_core::reactor::Core::run::{{closure}}::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.17/src/reactor/mod.rs:233
  59: tokio_executor::global::with_default::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.5/src/global.rs:192
  60: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  61: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  62: tokio_executor::global::with_default
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.5/src/global.rs:162
  63: tokio_core::reactor::Core::run::{{closure}}::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.17/src/reactor/mod.rs:232
  64: tokio_reactor::with_default::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.6/src/lib.rs:229
  65: <std::thread::local::LocalKey<T>>::try_with
             at libstd/thread/local.rs:294
  66: <std::thread::local::LocalKey<T>>::with
             at libstd/thread/local.rs:248
  67: tokio_reactor::with_default
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.6/src/lib.rs:212
  68: tokio_core::reactor::Core::run::{{closure}}
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.17/src/reactor/mod.rs:231
  69: <scoped_tls::ScopedKey<T>>::set
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-0.1.2/src/lib.rs:155
  70: tokio_core::reactor::Core::run
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-core-0.1.17/src/reactor/mod.rs:230
  71: irc::client::reactor::IrcReactor::run
             at /home/ed/.cargo/registry/src/github.com-1ecc6299db9ec823/irc-0.13.6/src/client/reactor.rs:190
  72: url_bot::main
             at src/main.rs:107
  73: std::rt::lang_start::{{closure}}
             at libstd/rt.rs:74
  74: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  75: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  76: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  77: std::rt::lang_start
             at libstd/rt.rs:74
  78: main
  79: __libc_start_main
  80: _start
```

**Comment by nuxeh** [comment]
After some investigation, it seems that this might be an issue with upstream `reqwest`, since `^` is never valid in a URL according to the URL specification https://www.w3.org/Addressing/URL/url-spec.txt

I have created an issue on `reqwest`:
https://github.com/seanmonstar/reqwest/issues/442

Also, this seems to be a separate issue from #82 , which appears to have been fixed by updating `reqwest` to `0.9.8`
